### PR TITLE
Remove email step

### PR DESCRIPTION
**Pull request by polmiro**
## Motivation / Description

Merges the email entry page with the payment entry page.

## Changes

- Uses a `LinkAuthenticationEmail` to capture the email with a stripe looking form input.
- Splits out integration tests.
- Adds test helpers for integration tests.
- Adds more integration tests

### CODE REVIEW

File: examples/webbilling-demo/src/tests/main.test.ts [hunk 1]
```diff
@@ -8,19 +8,27 @@ const RC_PAYWALL_TEST_OFFERING_ID = "rc_paywalls_e2e_test_2";
 const RC_PAYWALL_TEST_OFFERING_ID_WITH_VARIABLES =
   "rc_paywalls_e2e_test_variables_2";
 
-const waitForCheckoutStartRequest = (page: Page) => {
-  return page.waitForRequest(
-    (request) =>
-      request.url().includes("checkout/start") && request.method() === "POST",
-  );
+const waitForCheckoutStartRequest = (
```
Comment (vicfergar) [inline examples/webbilling-demo/src/tests/main.test.ts:11]: Does it make sense moving this helper function to the bottom as the others?

File: examples/webbilling-demo/src/tests/main.test.ts [hunk 2]
```diff
@@ -66,56 +72,108 @@ test.describe("Main", () => {
...
+    const userId = `${getUserId(browserName)}_subscription`;
+    const email = getEmailFromUserId(userId);
+    const page = await setupTest(browser, userId);
 
-    // Try with an invalid card declined server side
-    await enterCreditCardDetailsAndContinue(page, "4000003800000446");
+    const packageCards = await getPackageCards(page);
+    await startPurchaseFlow(packageCards[1]);
+    await enterEmail(page, email);
+    await enterCreditCardDetails(page, "4000003800000446");
```
Comment (vicfergar) [inline examples/webbilling-demo/src/tests/main.test.ts:173]:
```suggestion
    await enterCreditCardDetails(page, "4000 0038 0000 0446");
```

File: src/behavioural-events/sdk-event-helpers.ts [hunk 3]
```diff
@@ -33,15 +31,15 @@ export function createCheckoutFlowErrorEvent({
...
   errorCode,
   errorMessage,
 }: {
   errorCode: string | null;
   errorMessage: string;
-}): CheckoutBillingFormErrorEvent {
+}): CheckoutPaymentFormErrorEvent {
   return {
-    eventName: SDKEventName.CheckoutBillingFormError,
+    eventName: SDKEventName.CheckoutPaymentFormError,
```
Comment (vicfergar) [inline src/behavioural-events/sdk-event-helpers.ts:42]: Are we deprecating the `CheckoutBillingFormError` event type in favor of `CheckoutPaymentFormError`?

File: src/ui/localization/locale/en.json [hunk 4]
```diff
@@ -54,6 +54,7 @@
   "error_page.error_message_already_subscribed": "You are already subscribed to this product.",
   "error_page.error_message_already_purchased": "You have already purchased this product.",
   "error_page.error_message_missing_email_error": "Email is required to complete the purchase.",
```
Comment (vicfergar) [inline src/ui/localization/locale/en.json:56]: I guess this one has been replaced by a Stripe error message, right?

File: src/ui/molecules/pricing-table.svelte [hunk 5]
```diff
@@ -163,6 +163,10 @@
     color: var(--rc-color-grey-text-dark);
   }
 
+  .rcb-pricing-table-value {
+    text-align: right;
```
Comment (vicfergar) [inline src/ui/molecules/pricing-table.svelte:167]: This seems unrelated. Did you find a style issue in the pricing table?

File: src/ui/pages/payment-entry-page.svelte [hunk 6]
```diff
@@ -101,17 +135,36 @@
     await stripeSubmit();
   }
 
+  function handlePurchaseFlowError() {
```
Comment (vicfergar) [inline src/ui/pages/payment-entry-page.svelte:138]: It looks like this function is handling the dismissal of the error
```suggestion
  function handleDismissPurchaseFlowError() {
```

File: src/ui/pages/payment-entry-page.svelte [hunk 7]
```diff
@@ -150,27 +203,38 @@
...
 </script>
 
+{#if emailError}
+  <ErrorPage
+    lastError={emailError}
+    {productDetails}
+    supportEmail={brandingInfo?.support_email ?? null}
+    onContinue={handlePurchaseFlowError}
+  />
+{/if}
```
Comment (vicfergar) [inline src/ui/pages/payment-entry-page.svelte:213]: It’s a bit odd that we have different ways of representing `emailError`s and `modalErrorMessage`s. Is there a possibility to unify them?

File: src/ui/purchases-ui.svelte [hunk 8]
```diff
@@ -146,7 +131,7 @@
         productId,
         purchaseOption,
         rcPackage.webBillingProduct.presentedOfferingContext,
-        customerEmail,
+        undefined,
```
Comment (vicfergar) [inline src/ui/purchases-ui.svelte:134]: We can even remove this parameter from this helper method

File: src/ui/localization/locale/en.json (hunk 4 above)
Comment (polmiro) [reply src/ui/localization/locale/en.json:56]: Yep, there's still a bit of a mess with error codes for the public SDK getting tangled with the error codes needed by the UI but I decided to not fix it here.

File: src/ui/pages/payment-entry-page.svelte (hunk 7 above)
Comment (polmiro) [reply src/ui/pages/payment-entry-page.svelte:213]: Yes I agree. But I am not sure I want to increase the scope of the PR and this is a pre-existing issue.

File: src/behavioural-events/sdk-event-helpers.ts (hunk 3 above)
Comment (polmiro) [reply src/behavioural-events/sdk-event-helpers.ts:42]: The name convention indicates the page. Given that there is no billing form anymore I prefer to use `CheckoutPaymentFormError`.

File: src/ui/molecules/pricing-table.svelte (hunk 5 above)
Comment (polmiro) [reply src/ui/molecules/pricing-table.svelte:167]: I have no idea how this got here 🤔 . Thanks for the heads up.

File: examples/webbilling-demo/src/tests/main.test.ts (hunk 1 above)
Comment (polmiro) [reply examples/webbilling-demo/src/tests/main.test.ts:11]: On it.

File: src/ui/purchases-ui.svelte (hunk 8 above)
Comment (polmiro) [reply src/ui/purchases-ui.svelte:134]: This just opened up some questions on my end:

1. What happens if the user already has an entitlement and it is configured not allow more than one purchase? The integration test is relying on a mock of the request which is just shooting ourselves in the foot I think.
2. I made a change (I was just confirming with Ed) that shows the email input prefilled with the email passed by the developer in the query parameters. This seems to conflict with the entitlement check described above. I may have to undo it.

Comment (polmiro) [reply src/ui/purchases-ui.svelte:134]: Talked to Ed, going to keep the existing behavior so I will rework to keep this the way it was when an email is passed via query parameters which will mean this field will actually have a value passed on it.

File: src/ui/pages/payment-entry-page.svelte (hunk 7 above)
Comment (polmiro) [reply src/ui/pages/payment-entry-page.svelte]: I am actually having to come back to this since I am discovering some edge cases as I am adding more integration tests.

File: src/ui/purchases-ui.svelte [hunk 9]
```diff
@@ -218,14 +212,23 @@
     if (taxCalculation.error) {
       switch (taxCalculation.error) {
         case TaxCalculationError.Pending:
-          priceBreakdown.taxCalculationStatus = "pending";
+          priceBreakdown = {
+            ...priceBreakdown,
+            taxCalculationStatus: "pending",
```
Comment (vicfergar) [inline src/ui/purchases-ui.svelte:217]:
```suggestion
            taxCalculationStatus: "pending",
            pendingReason: null,
```

File: src/ui/theme/spacing.ts [hunk 10]
```diff
@@ -52,4 +52,9 @@ export const DEFAULT_SPACING: Spacing = {
     tablet: "3rem",
     desktop: "3rem",
   },
+  gapStripeElement: {
```
Comment (polmiro) [inline src/ui/theme/spacing.ts:55]: @alfondotnet I did this to match a bit more closer the spacing between the `LinkAuthenticationStripeElement` and the `PaymentElement` to the input elements within the `PaymentElement` from Stripe. Not super elegant but I guess it's fine?

File: examples/webbilling-demo/src/util/PurchasesLoader.ts [hunk 11]
```diff
@@ -37,7 +37,11 @@ const loadPurchases: LoaderFunction<IPurchasesLoaderData> = async ({
       Purchases.configure(
         apiKey,
         appUserId,
-        {},
+        {
+          additionalHeaders: {
+            "X-Rc-Canary": "noemail",
+          },
+        },
```
Comment (vicfergar) [inline examples/webbilling-demo/src/util/PurchasesLoader.ts:44]: This was not intentional, right?
```suggestion
        {},
```

File: src/behavioural-events/sdk-event-helpers.ts [hunk 12]
```diff
@@ -6,12 +6,10 @@ import type {
...
   CheckoutPaymentFormSubmitEvent,
+  CheckoutPaymentFormErrorEvent,
+  CheckoutSessionStartEvent,
 } from "./sdk-events";
-import {
-  SDKEventName,
-  type CheckoutBillingFormErrorEvent,
-  type CheckoutSessionStartEvent,
-} from "./sdk-events";
+import { SDKEventName } from "./sdk-events";
```
Comment (vicfergar) [inline src/behavioural-events/sdk-event-helpers.ts:12]: Why is `SDKEventName` not also merged in the previous `import`?

File: src/behavioural-events/sdk-events.ts [hunk 13]
```diff
@@ -128,6 +130,14 @@ export interface CheckoutPaymentFormImpressionEvent extends ISDKEvent {
   eventName: SDKEventName.CheckoutPaymentFormImpression;
 }
 
+export interface CheckoutPaymentFormErrorEvent extends ISDKEvent {
```
Comment (vicfergar) [inline src/behavioural-events/sdk-events.ts:133]: Should we get rid of the `CheckoutBillingForm*` events?

File: src/networking/responses/checkout-complete-response.ts [hunk 14]
```diff
@@ -1,6 +1,6 @@
 export interface CheckoutCompleteResponse {
   operation_session_id: string;
   gateway_params: {
-    client_secret?: string;
+    client_secret?: string | null;
```
Comment (vicfergar) [inline src/networking/responses/checkout-complete-response.ts:4]: Is this needed? it was already `undefined`, right?

File: src/ui/localization/supportedLanguages.ts [hunk 15]
```diff
@@ -88,6 +88,7 @@ export enum LocalizationKeys {
   ErrorPageErrorMessageAlreadySubscribed = "error_page.error_message_already_subscribed",
   ErrorPageErrorMessageAlreadyPurchased = "error_page.error_message_already_purchased",
   ErrorPageErrorMessageMissingEmailError = "error_page.error_message_missing_email_error",
+  ErrorPageErrorMessageInvalidEmailError = "error_page.error_message_invalid_email_error",
```
Comment (vicfergar) [inline src/ui/localization/supportedLanguages.ts:91]: Could we remove the `email_entry_page.*` keys?

File: src/ui/molecules/stripe-authentication-link-element.svelte [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    event: StripeLinkAuthenticationElementChangeEvent,
+  ) => void | Promise<void>;
+  export let onError: undefined | ((error: any) => void | Promise<void>) =
+    undefined;
+  export let onReady: undefined | (() => void | Promise<void>) = undefined;
+  export let email: string | undefined;
+  export let elements: StripeElements;
+
+  let linkAuthenticationElement: StripeLinkAuthenticationElement | null = null;
+  let linkAuthenticationElementId = "link-authentication-element";
```
Comment (vicfergar) [inline src/ui/molecules/stripe-authentication-link-element.svelte:21]:
```suggestion
  const linkAuthenticationElementId = "link-authentication-element";
```

File: src/stories/pages/purchase.stories.svelte [hunk 17]
```diff
@@ -66,6 +66,7 @@
     {brandingInfo}
     onContinue={() => {}}
     closeWithError={() => {}}
+    customerEmail={args.customerEmail}
```
Comment (vicfergar) [inline src/stories/pages/purchase.stories.svelte:69]: Could we add some stories with this field set?

File: src/ui/purchases-ui.svelte [hunk 18]
```diff
@@ -149,31 +152,32 @@
...
+            appUserId,
+            productId,
+            purchaseOption,
+            rcPackage.webBillingProduct.presentedOfferingContext,
+            email,
+            metadata,
+          );
+        } else {
+          throw e;
+        }
```
Comment (vicfergar) [inline src/ui/purchases-ui.svelte:179]: Could we cover this in a test?

File: examples/webbilling-demo/src/util/PurchasesLoader.ts (hunk 11 above)
Comment (polmiro) [reply examples/webbilling-demo/src/util/PurchasesLoader.ts]: Thanks, it got late 😓

File: src/behavioural-events/sdk-events.ts (hunk 13 above)
Comment (polmiro) [reply src/behavioural-events/sdk-events.ts:133]: I guess so 👍🏻

File: src/networking/responses/checkout-complete-response.ts (hunk 14 above)
Comment (polmiro) [reply src/networking/responses/checkout-complete-response.ts:4]: I think I will leave it as is. I can confirm the backend returns null but not sure if removing the undefined could lead to any cases I am not seeing right now.

**Review by vicfergar — APPROVED** [review]
LG! Thanks for addressing all the feedback.
